src/devices/sound/qsoundhle.cpp

```diff
@@ -180,70 +212,429 @@ WRITE8_MEMBER(qsound_hle_device::qsound_w)
 
 READ8_MEMBER(qsound_hle_device::qsound_r)
 {
-	/* Port ready bit (0x80 if ready) */
-	return 0x80;
+	// ready bit (0x00 = busy, 0x80 == ready)
+	m_stream->update();
+	return m_ready_flag;
 }
 
 
 void qsound_hle_device::write_data(u8 address, u16 data)
 {
-	int ch = 0, reg;
+	uint16_t *destination = register_map[address];
+	if(destination)
+		*destination = data;
+	m_ready_flag = 0;
+}
 
-	if (address < 0x80)
+void qsound_hle_device::init_register_map()
+{
+	// unused registers
+	for(int i=0;i<256;i++)
+		register_map[i] = NULL;
+
+	// PCM registers
+	for(int i=0;i<16;i++) // PCM voices
 	{
-		ch = address >> 3;
-		reg = address & 7;
+		register_map[(i<<3)+0] = (uint16_t*)&m_voice[(i+1)%16].bank; // Bank applies to the next channel
+		register_map[(i<<3)+1] = (uint16_t*)&m_voice[i].addr; // Current sample position and start position.
+		register_map[(i<<3)+2] = (uint16_t*)&m_voice[i].rate; // 4.12 fixed point decimal.
+		register_map[(i<<3)+3] = (uint16_t*)&m_voice[i].phase;
+		register_map[(i<<3)+4] = (uint16_t*)&m_voice[i].loop_len;
+		register_map[(i<<3)+5] = (uint16_t*)&m_voice[i].end_addr;
+		register_map[(i<<3)+6] = (uint16_t*)&m_voice[i].volume;
+		register_map[(i<<3)+7] = NULL;	// unused
+		register_map[i+0x80] = (uint16_t*)&m_voice_pan[i];
+		register_map[i+0xba] = (uint16_t*)&m_voice[i].echo;
 	}
-	else if (address < 0x90)
+
+	// ADPCM registers
+	for(int i=0;i<3;i++) // ADPCM voices
 	{
-		ch = address & 0xf;
-		reg = 8;
+		// ADPCM sample rate is fixed to 8khz. (one channel is updated every third sample)
+		register_map[(i<<2)+0xca] = (uint16_t*)&m_adpcm[i].start_addr;
+		register_map[(i<<2)+0xcb] = (uint16_t*)&m_adpcm[i].end_addr;
+		register_map[(i<<2)+0xcc] = (uint16_t*)&m_adpcm[i].bank;
+		register_map[(i<<2)+0xcd] = (uint16_t*)&m_adpcm[i].volume;
+		register_map[i+0xd6] = (uint16_t*)&m_adpcm[i].flag; // non-zero to start ADPCM playback
+		register_map[i+0x90] = (uint16_t*)&m_voice_pan[16+i];
 	}
-	else if (address >= 0xba && address < 0xca)
+
+	// QSound registers
+	register_map[0x93] = (uint16_t*)&m_echo.feedback;
+	register_map[0xd9] = (uint16_t*)&m_echo.end_pos;
+	register_map[0xe2] = (uint16_t*)&m_delay_update; // non-zero to update delays
+	register_map[0xe3] = (uint16_t*)&m_next_state;
+	for(int i=0;i<2;i++)  // left, right
 	{
-		ch = address - 0xba;
-		reg = 9;
+		// Wet
+		register_map[(i<<1)+0xda] = (uint16_t*)&m_filter[i].table_pos;
+		register_map[(i<<1)+0xde] = (uint16_t*)&m_wet[i].delay;
+		register_map[(i<<1)+0xe4] = (uint16_t*)&m_wet[i].volume;
+		// Dry
+		register_map[(i<<1)+0xdb] = (uint16_t*)&m_alt_filter[i].table_pos;
+		register_map[(i<<1)+0xdf] = (uint16_t*)&m_dry[i].delay;
+		register_map[(i<<1)+0xe5] = (uint16_t*)&m_dry[i].volume;
+	}
+}
+
+int16_t qsound_hle_device::get_sample(uint16_t bank,uint16_t address)
+{
+	uint32_t rom_addr;
+	uint8_t sample_data;
```

Comment:
In cases like this, please declare variables at initialisation and use `const` where appropriate to reduce mutable state.